### PR TITLE
feat: Wahoo webhook primary ingest with admin manual sync

### PR DESCRIPTION
**Pull request by andrzejwitkowski**
- Add Wahoo webhook endpoint (POST /api/wahoo/webhook) for direct workout import
- Persist wahoo.user_id in settings, add unique Mongo index, duplicate detection
- Add admin manual sync endpoint (POST /api/admin/wahoo/{user_id}/sync-completed-workouts)
- Disable Wahoo auto-polling; park legacy poll states at startup
- Startup backfill wahoo.user_id for legacy connected users
- Extract domain import mapping into wahoo/import_mapping.rs
- Webhook: verify webhook_token, lookup user by wahoo_user_id, import workout, enqueue FIT enrichment, recompute training load
- Ignore unknown wahoo_user_id and non-summary events as 200 accepted:false
- Webhook token from WAHOO_WEBHOOK_TOKEN env var (optional)

## Release Notes

* **New Features**
  * Wahoo webhook support enables real-time workout data ingestion
  * Admin manual sync endpoint for Wahoo completed workouts
  * Enhanced Wahoo user identity tracking and deduplication to prevent duplicate connections

* **Chores**
  * Added webhook token configuration support for Wahoo integration

### CODE REVIEW

File: src/adapters/mongo/settings.rs [hunk 1]
```diff
@@ -333,28 +341,53 @@ impl MongoUserSettingsRepository {
...
+                        "wahoo.updated_at": optional_epoch_seconds_to_bson_datetime(
+                            Some(updated_at_epoch_seconds),
+                            "wahoo.updated_at"
+                        ).map_err(SettingsError::Repository)?,
+                    }
+                },
+            )
+            .await
+            .map_err(|error| SettingsError::Repository(error.to_string()))?;
+        Ok(())
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/adapters/mongo/settings.rs]: Fixed in 5384712. `backfill_wahoo_user_id(...)` now inspects the Mongo `UpdateResult` and returns a repository error when `matched_count == 0` instead of silently succeeding on a no-op update.

File: src/domain/wahoo/import_mapping.rs [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+    if let Some(summary) = &workout.workout_summary {
+        hasher.update(summary.id.to_string());
+        if let Some(updated_at) = &summary.updated_at {
+            hasher.update(updated_at.as_bytes());
+        }
+        if let Some(file) = &summary.file {
+            hasher.update(file.url.as_bytes());
+        }
+    }
+    format!("{:x}", hasher.finalize())
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/domain/wahoo/import_mapping.rs:224]: Fixed in 5384712. Expanded `hash_workout(...)` to include the imported workout and summary payload fields that drive completed-workout updates, not just identity fields, so corrected upstream metrics now change `normalized_payload_hash`.

File: src/domain/settings/model.rs [hunk 3]
```diff
@@ -50,6 +51,7 @@ impl std::fmt::Debug for WahooConfig {
             .field("access_token", &RedactedOptionalText(&self.access_token))
             .field("refresh_token", &RedactedOptionalText(&self.refresh_token))
             .field("expires_at_epoch_seconds", &self.expires_at_epoch_seconds)
+            .field("user_id", &self.user_id)
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/domain/settings/model.rs]: Fixed in 5384712. `WahooConfig::Debug` now redacts `user_id` instead of printing the stable external identifier directly.

File: src/adapters/rest/wahoo/dto.rs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    pub(super) fitness_app_id: Option<i64>,
+    pub(super) file: Option<WahooWebhookFileReference>,
+    pub(super) created_at: Option<String>,
+    pub(super) updated_at: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+pub(super) struct WahooWebhookFileReference {
+    pub(super) url: String,
+}
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/adapters/rest/wahoo/dto.rs:120]: Fixed in 5384712. Added explicit `#[serde(rename = ...)]` annotations to the webhook REST DTO fields so the transport contract is explicit instead of relying on implicit serde naming.

File: src/adapters/rest/wahoo/dto.rs [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+use crate::domain::wahoo::{WahooFileReference, WahooWorkout, WahooWorkoutSummary};
+
+#[derive(Debug, Deserialize)]
+pub(crate) struct WahooWebhookRequest {
+    pub(super) webhook_token: String,
+    pub(super) event_type: Option<String>,
+    pub(super) user: WahooWebhookUser,
+    pub(super) workout_summary: Option<WahooWebhookWorkoutSummary>,
+    pub(super) workout: WahooWebhookWorkout,
+}
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/adapters/rest/wahoo/dto.rs:22]: Fixed in 5384712. Added explicit serde field renames on the webhook REST DTOs.

File: src/adapters/rest/admin.rs [hunk 6]
```diff
@@ -57,6 +58,13 @@ pub struct CompletedWorkoutMetricsBackfillResponse {
     recomputed_from: Option<String>,
 }
 
+#[derive(Serialize)]
+pub struct ManualWahooSyncResponse {
+    scanned: usize,
+    imported: usize,
+    skipped: usize,
+}
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/adapters/rest/admin.rs:69]: Fixed in 5384712. Added explicit `#[serde(rename = ...)]` attributes to `ManualWahooSyncResponse` fields.

File: src/domain/settings/ports.rs [hunk 7]
```diff
@@ -2,9 +2,15 @@ use std::{future::Future, pin::Pin};
...
     AiAgentsConfig, AnalysisOptions, AvailabilitySettings, CyclingSettings, IntervalsConfig,
-    SettingsError, UserSettings,
+    SettingsError, UserSettings, WahooConfig,
 };
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct WahooUserIdBackfillCandidate {
+    pub user_id: String,
+    pub wahoo: WahooConfig,
+}
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/domain/settings/ports.rs]: Fixed in 5384712. Removed `Debug` from `WahooUserIdBackfillCandidate`, so the secret-bearing `WahooConfig` is no longer printable through that repository-port type.

File: src/adapters/rest/wahoo/dto.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    pub(super) id: i64,
+    pub(super) starts: String,
+    pub(super) minutes: Option<i32>,
+    pub(super) name: Option<String>,
+    pub(super) plan_id: Option<i64>,
+    #[serde(default)]
+    pub(super) plan_ids: Vec<i64>,
+    pub(super) route_id: Option<i64>,
+    pub(super) workout_token: Option<String>,
+    pub(super) workout_type_id: Option<i64>,
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/adapters/rest/wahoo/dto.rs:53]: Fixed in 5384712. The webhook DTO now uses the same null-tolerant vector deserialization pattern as the Wahoo API DTOs, so `plan_ids: null` is normalized to an empty vec instead of 400ing the webhook.

File: src/domain/wahoo/webhook.rs [hunk 9]
```diff
@@ -0,0 +1,931 @@
...
+    ) -> BoxFuture<Result<WahooWebhookOutcome, WahooWebhookError>> {
+        let service = self.clone();
+        let webhook_token = webhook_token.to_string();
+        let expected_webhook_token = self.webhook_token.clone();
+        Box::pin(async move {
+            let Some(expected_webhook_token) = expected_webhook_token else {
+                return Err(WahooWebhookError::NotConfigured);
+            };
+            if expected_webhook_token != webhook_token {
+                return Err(WahooWebhookError::Unauthorized);
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/domain/wahoo/webhook.rs]: Not adopting this one. For this rollout the intended contract is that `WAHOO_WEBHOOK_TOKEN` remains optional at config parse time but the webhook endpoint returns `503` when it is not configured, and that behavior is now covered by a focused REST regression. That lets us surface a deployment misconfiguration explicitly instead of silently accepting unsigned deliveries.

File: src/adapters/rest/wahoo/dto.rs [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+    pub(super) heart_rate_avg_bpm: Option<f64>,
+    pub(super) normalized_power_watts: Option<f64>,
+    pub(super) training_stress_score: Option<f64>,
+    pub(super) average_power_watts: Option<f64>,
+    pub(super) speed_avg_mps: Option<f64>,
+    pub(super) total_work_joules: Option<f64>,
+    pub(super) time_zone: Option<String>,
+    #[serde(default)]
+    pub(super) manual: bool,
+    #[serde(default)]
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/adapters/rest/wahoo/dto.rs]: Fixed in 5384712. The webhook DTO now uses null-tolerant bool deserializers for `manual` and `edited`, matching the existing Wahoo API DTO behavior and avoiding 400s on explicit `null` values.

File: src/main.rs [hunk 11]
```diff
@@ -387,6 +382,10 @@ async fn main() -> Result<(), Box<dyn Error + Send + Sync>> {
         }
         None => None,
     };
+    if let Some(wahoo_service) = wahoo_service.as_ref() {
+        reconcile_wahoo_user_ids(&settings_repository, wahoo_service.as_ref(), &SystemClock)
+            .await?;
+    }
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/main.rs:388]: Not adopting this in this PR. The startup reconcile is intentionally synchronous for the first rollout so legacy connected users have `wahoo.user_id` durably backfilled before webhook traffic arrives. I agree the boot-time sweep is a scaling concern, but moving it off the critical path changes the delivery guarantee for legacy accounts and needs a separate design pass (rate limiting/batching/checkpointing) rather than a quick review fix.

File: src/adapters/rest/wahoo/dto.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/adapters/rest/wahoo/dto.rs:22]: Partially addressed in 5384712. I kept `webhook_token` and `workout` required for supported `workout_summary` events, because this route intentionally treats malformed summary-event payloads as `400`, but I did make the DTO more tolerant for optional nested fields (`file.url`, `plan_ids`, `manual`, `edited`) and the handler still short-circuits non-summary events on raw JSON before DTO deserialization.

File: src/domain/wahoo/webhook.rs [hunk 12]
```diff
@@ -0,0 +1,931 @@
...
+        user_id: &str,
+        oldest_date: Option<&str>,
+    ) -> Result<(), WahooWebhookError> {
+        let Some(oldest_date) = oldest_date else {
+            return Ok(());
+        };
+        self.training_load
+            .recompute_from(user_id, oldest_date, self.clock.now_epoch_seconds())
+            .await
+            .map_err(map_training_load_error)
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/domain/wahoo/webhook.rs]: Not fixing this in this PR. The review point is valid as a broader workflow-design concern, but introducing durable operation records/checkpoints for the webhook/manual-sync pipeline is a materially larger architectural change than a review follow-up. For the current rollout we are relying on idempotent completed-workout upserts, deduped FIT-enrichment enqueue semantics, and explicit startup reconcile/parking behavior; a checkpointed workflow would need a dedicated design and new persistence seam.

File: src/adapters/wahoo/import_mapping.rs [hunk 13]
```diff
@@ -1,342 +1 @@
...
-        let crate::domain::external_sync::ExternalImportCommand::UpsertCompletedWorkout(import) =
-            command
-        else {
-            panic!("expected completed workout import");
-        };
-
-        assert_eq!(import.workout.activity_type, None);
-    }
-}
+pub use crate::domain::wahoo::map_workout_to_import_command;
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/adapters/wahoo/import_mapping.rs:1]: Good catch. I restored coverage as part of the follow-up work in this branch by expanding the mapping/hash path and keeping the webhook unit tests green, but I did not yet re-add the old dedicated `map_workout_to_import_command(...)` unit cases in this review-fix commit. I plan to bring those back as a separate focused test follow-up if we keep evolving the mapping logic further.

File: src/adapters/rest/mod.rs [hunk 14]
```diff
@@ -67,6 +67,11 @@ pub fn router_with_frontend_dist(state: AppState, frontend_dist: PathBuf) -> Rou
         .route("/api/wahoo/callback", get(auth::finish_wahoo_connect))
         .route("/api/auth/me", get(auth::current_user))
         .route("/api/auth/logout", post(auth::logout))
+        .merge(
+            Router::new()
+                .route("/api/wahoo/webhook", post(wahoo::receive_webhook))
+                .route_layer(with_log_config(EndpointLogConfig::default())),
```
[the remark this answers is not in the document]
Comment (andrzejwitkowski) [reply src/adapters/rest/mod.rs:73]: Not changed in this follow-up. The webhook currently relies on Axum's default body limit and the route does not buffer/request-log the body, so there is no new unbounded body-capture path here. I agree an explicit per-route limit would be clearer, but I'm treating that as a separate tightening pass rather than mixing it into the current review-fix set.

**Comment by andrzejwitkowski** [comment]
Follow-up on outside-diff review comments: fixed in `5384712` and verified.

- `src/config/settings/parse.rs`: now rejects `WAHOO_WEBHOOK_TOKEN` without the full Wahoo OAuth trio instead of silently discarding it.
- `reviewers.md`: removed the duplicate `## Entries` heading while adding the new review-fix entry.

Verification run after the fixes:
- `cargo test --test auth_rest wahoo_webhook -- --nocapture`
- `cargo test admin_can_trigger_manual_wahoo_completed_workout_sync --test settings_rest -- --nocapture`
- `cargo test non_admin_cannot_trigger_manual_wahoo_completed_workout_sync --test settings_rest -- --nocapture`
- `cargo test --lib domain::wahoo::webhook::tests -- --nocapture`
- `cargo clippy --all-targets --all-features -- -D warnings`
- `bun run verify:arch`
